### PR TITLE
chore(DHT): WS library client for NodeJS websocket client connections

### PR DESCRIPTION
**Pull request by juslesan**
## Summary

Use the ws library to open client ws connections in NodeJS. `websocket` library is still used in browsers to avoid compatibility issues with between different browsers WebSocket implementations.

This change improves the CPU use in NodeJS

Should fix this funky error in client e2e tests:
```
Test suite failed to run
    TypeError: require(...) is not a function
      at Object.createBufferingLogger [as BufferingLogger] (../../node_modules/websocket/lib/utils.js:24:39)
      at new WebSocketConnection (../../node_modules/websocket/lib/WebSocketConnection.js:42:25)
      at WebSocketClient.Object.<anonymous>.WebSocketClient.succeedHandshake (../../node_modules/websocket/lib/WebSocketClient.js:343:22)
      at WebSocketClient.Object.<anonymous>.WebSocketClient.validateHandshake (../../node_modules/websocket/lib/WebSocketClient.js:332:10)
      at ClientRequest.handleRequestUpgrade (../../node_modules/websocket/lib/WebSocketClient.js:2[61](https://github.com/streamr-dev/network/actions/runs/7972200333/job/21763492753?pr=2383#step:8:62):14)
```

## Other changes

- Renamed `ClientWebsocket` -> `WebsocketClientConnection` (BrowserWebsocketClientConnection for browsers)

### CODE REVIEW

File: packages/dht/src/connection/websocket/WebsocketClientConnection.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import { WebSocket } from 'ws'
+import { ConnectionEvents, ConnectionID, ConnectionType, IConnection } from '../IConnection'
+import { createRandomConnectionId } from '../Connection'
+import { CUSTOM_GOING_AWAY, GOING_AWAY } from './WebsocketConnector'
+
+const logger = new Logger(module)
+
+const BINARY_TYPE = 'nodebuffer'
+
+export class WebsocketClientConnection extends EventEmitter<ConnectionEvents> implements IConnection {
```
Comment (harbu) [inline packages/dht/src/connection/websocket/WebsocketClientConnection.ts]: Lots of duplicated code / logic between this and `BrowserWebsocketClientConnection`.

I'd recommend moving the shared code into an abstract base class and then having this class and `BrowserWebsocketClientConnection` only implement their respective library specific logic.

File: packages/dht/src/connection/websocket/BrowserWebsocketClientConnection.ts [hunk 2]
```diff
@@ -3,24 +3,17 @@ import EventEmitter from 'eventemitter3'
 import { ICloseEvent, IMessageEvent, w3cwebsocket as Websocket } from 'websocket'
 import { ConnectionEvents, ConnectionID, ConnectionType, IConnection } from '../IConnection'
 import { createRandomConnectionId } from '../Connection'
+import { CUSTOM_GOING_AWAY, GOING_AWAY } from './WebsocketConnector'
```
Comment (ptesavol) [inline packages/dht/src/connection/websocket/BrowserWebsocketClientConnection.ts]: Isn't this a circular import?

**Review by ptesavol — APPROVED** [review]
Approved, one question in comments.

**Review by harbu — CHANGES_REQUESTED** [review]

File: packages/dht/src/connection/websocket/WebsocketClientConnection.ts (hunk 1 above)
Comment (juslesan) [reply packages/dht/src/connection/websocket/WebsocketClientConnection.ts]: Added AbstractWebsocketClientConnection

**Review by harbu — APPROVED** [review]

**Comment by juslesan** [comment]
There seem to be problems with establishing hundreds connections with ws library's client. So far this can has only been replicated in localhost. Could be that it is not a real issue in production but it has to be validated before merge.

File: packages/dht/src/connection/websocket/AbstractWebsocketClientConnection.ts [hunk 3]
```diff
@@ -1,89 +1,47 @@
...
-
-    private doDisconnect(code?: number, reason?: string) {
-        this.destroyed = true
-        this.stopListening()
-        this.socket = undefined
-        const gracefulLeave = (code === GOING_AWAY) || (code === CUSTOM_GOING_AWAY)
-        this.emit('disconnected', gracefulLeave, code, reason)
-        this.removeAllListeners()
-    }
+    public abstract connect(address: string, selfSigned?: boolean): void
```
Comment (teogeb) [inline packages/dht/src/connection/websocket/AbstractWebsocketClientConnection.ts:37]: Keep the comment:
`"// TODO explicit default value for "selfSigned" or make it required"`

File: packages/dht/src/connection/websocket/BrowserWebsocketClientConnection.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        if (!this.destroyed) {
+            this.socket = new Websocket(address, undefined, undefined, undefined, { rejectUnauthorized: !selfSigned })
+            this.socket.binaryType = BINARY_TYPE
+            this.socket.onerror = (error: Error) => this.onError(error)
+            this.socket.onopen = () => this.onOpen()
+            this.socket.onclose = (event: ICloseEvent) => this.onClose(event.code, event.reason)
+            this.socket.onmessage = (message: IMessageEvent) => {
+                if (!this.destroyed) {
+                    if (typeof message.data === 'string') {
+                        logger.debug('Received string: \'' + message.data + '\'')
```
Comment (teogeb) [inline packages/dht/src/connection/websocket/BrowserWebsocketClientConnection.ts]: This in unsupported feature? Could mention that in the log message (also in `WebsocketClientConnection`)

File: packages/dht/tsconfig.browser.json [hunk 5]
```diff
@@ -11,6 +11,7 @@
     "package.json"
   ],
   "exclude": [
-    "src/connection/webrtc/NodeWebrtcConnection.ts"
+    "src/connection/webrtc/NodeWebrtcConnection.ts",
+    "src/connection/websocket/NodeWebSocketClientConnection.ts"
```
Comment (teogeb) [inline packages/dht/tsconfig.browser.json]:
```suggestion
    "src/connection/websocket/NodeWebsocketClientConnection.ts"
```

File: packages/dht/tsconfig.jest.json [hunk 6]
```diff
@@ -10,7 +10,8 @@
         "package.json"
     ],
     "exclude": [
-        "src/connection/webrtc/BrowserWebrtcConnection.ts"
+        "src/connection/webrtc/BrowserWebrtcConnection.ts",
+        "src/connection/websocket/BrowserWebSocketClientConnection.ts"
```
Comment (teogeb) [inline packages/dht/tsconfig.jest.json]:
```suggestion
        "src/connection/websocket/BrowserWebsocketClientConnection.ts"
```

**Review by teogeb — CHANGES_REQUESTED** [review]

File: packages/dht/tsconfig.node.json [hunk 7]
```diff
@@ -9,7 +9,8 @@
         "package.json"
     ],
     "exclude": [
-        "src/connection/webrtc/BrowserWebrtcConnection.ts"
+        "src/connection/webrtc/BrowserWebrtcConnection.ts",
+        "src/connection/websocket/BrowserWebSocketClientConnection.ts"
```
Comment (teogeb) [inline packages/dht/tsconfig.node.json]:
```suggestion
        "src/connection/websocket/BrowserWebsocketClientConnection.ts"
```

File: packages/dht/src/connection/websocket/AbstractWebsocketClientConnection.ts (hunk 3 above)
Comment (juslesan) [reply packages/dht/src/connection/websocket/AbstractWebsocketClientConnection.ts:37]: https://github.com/streamr-dev/network/pull/2092/files

**Review by teogeb — APPROVED** [review]

**Pull request merged**